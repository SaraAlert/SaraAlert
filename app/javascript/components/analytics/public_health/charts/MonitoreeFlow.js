import React from 'react';
import { PropTypes } from 'prop-types';
import { Card, Col, Row } from 'react-bootstrap';
import { formatPercentage } from '../../../../utils/Analytics';

const RELEASE_DATE = '10/5/2021';
const WORKFLOWS = ['Exposure', 'Isolation'];
const MONITOREE_FLOW_HEADERS = ['Last 24 Hours', 'Last 7 Days', 'Last 14 Days', 'Total'];

class MonitoreeFlow extends React.Component {
  constructor(props) {
    super(props);
    this.workflowTableData = WORKFLOWS.map(workflow => {
      return MONITOREE_FLOW_HEADERS.map(time_frame => {
        let thisTimeFrameData = props.stats.monitoree_snapshots.find(
          monitoree_snapshot => monitoree_snapshot.status === workflow && monitoree_snapshot.time_frame === time_frame
        );
        let inTotal = thisTimeFrameData?.new_enrollments + thisTimeFrameData?.transferred_in;
        let outTotal = thisTimeFrameData?.closed + thisTimeFrameData?.transferred_out;
        return {
          time_frame,
          new_enrollments: {
            value: thisTimeFrameData?.new_enrollments || 0,
            percentage: formatPercentage(thisTimeFrameData?.new_enrollments, inTotal),
          },
          transferred_in: {
            value: thisTimeFrameData?.transferred_in || 0,
            percentage: formatPercentage(thisTimeFrameData?.transferred_in, inTotal),
          },
          closed: {
            value: thisTimeFrameData?.closed || 0,
            percentage: formatPercentage(thisTimeFrameData?.closed, outTotal),
          },
          transferred_out: {
            value: thisTimeFrameData?.transferred_out || 0,
            percentage: formatPercentage(thisTimeFrameData?.transferred_out, outTotal),
          },
        };
      });
    });

    this.exposureToCaseTableData = MONITOREE_FLOW_HEADERS.map(time_frame => {
      let thisTimeFrameData = props.stats.monitoree_snapshots.find(
        monitoree_snapshot => monitoree_snapshot.status === 'Isolation' && monitoree_snapshot.time_frame === time_frame
      );
      return {
        time_frame,
        exposure_to_isolation_active: {
          value: thisTimeFrameData?.exposure_to_isolation_active || 0,
          percentage: null, // ADD ME
        },
        exposure_to_isolation_not_active: {
          value: thisTimeFrameData?.exposure_to_isolation_not_active || 0,
          percentage: null, // ADD ME
        },
        exposure_to_isolation_closed_in_exposure: {
          value: thisTimeFrameData?.exposure_to_isolation_closed_in_exposure || 0,
          percentage: null, // ADD ME
        },
        exposure_to_isolation_total: {
          value: thisTimeFrameData?.exposure_to_isolation_total || 0,
          percentage: null, // ADD ME
        },
      };
    });

    this.isolationToExposureTableData = MONITOREE_FLOW_HEADERS.map(time_frame => {
      let thisTimeFrameData = props.stats.monitoree_snapshots.find(
        monitoree_snapshot => monitoree_snapshot.status === 'Exposure' && monitoree_snapshot.time_frame === time_frame
      );
      return {
        time_frame,
        isolation_to_exposure_total: {
          value: thisTimeFrameData?.isolation_to_exposure_total || 0,
          percentage: null, // add me
        },
      };
    });
  }

  renderWorkflowTable(data, index) {
    return (
      <Col xl="12" key={index}>
        <div className="analytics-table-header">{WORKFLOWS[Number(index)]} Workflow</div>
        <div className="table-responsive">
          <table className="analytics-table">
            <thead>
              <tr className="g-border-bottom text-center header">
                <th></th>
                <th>Last 24h</th>
                <th>Last 7d</th>
                <th>Last 14d</th>
                <th>Cumulative</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td className="header" colSpan="5">
                  Incoming
                </td>
              </tr>
              <tr>
                <td className="sub-header indent">New Enrollments</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.new_enrollments.value}</span>
                      <span className="percentage align-bottom">({x.new_enrollments.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
              <tr className="g-border-bottom">
                <td className="sub-header indent">Transferred In</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.transferred_in.value}</span>
                      <span className="percentage">({x.transferred_in.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
              <tr>
                <td className="header" colSpan="5">
                  Outgoing
                </td>
              </tr>
              <tr>
                <td className="sub-header indent">Closed</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.closed.value}</span>
                      <span className="percentage">({x.closed.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
              <tr className="g-border-bottom">
                <td className="sub-header indent">Transferred Out</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.transferred_out.value}</span>
                      <span className="percentage">({x.transferred_out.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
            </tbody>
          </table>
        </div>
      </Col>
    );
  }

  renderExposureToCaseTable(data) {
    return (
      <Col xl="12">
        <div className="analytics-table-header">Exposure to Case Development</div>
        <div className="table-responsive">
          <table className="analytics-table">
            <thead>
              <tr className="g-border-bottom text-center header">
                <th></th>
                <th>Last 24h</th>
                <th>Last 7d</th>
                <th>Last 14d</th>
                <th>Cumulative</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td className="header" colSpan="5">
                  Moved from Exposure to Isolation Workflow
                </td>
              </tr>
              <tr>
                <td className="sub-header indent">Currently Active</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.exposure_to_isolation_active.value}</span>
                      <span className="percentage align-bottom">({x.exposure_to_isolation_active.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
              <tr className="g-border-bottom">
                <td className="sub-header indent">Closed or Purged</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.exposure_to_isolation_not_active.value}</span>
                      <span className="percentage">({x.exposure_to_isolation_not_active.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
              <tr className="b-border-bottom">
                <td className="header">Cases that were closed in Exposure Workflow</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.exposure_to_isolation_closed_in_exposure.value}</span>
                      <span className="percentage">({x.exposure_to_isolation_closed_in_exposure.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
              <tr className="g-border-bottom">
                <td className="header">Total Contacts that became Cases</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.exposure_to_isolation_total.value}</span>
                      <span className="percentage"></span>
                    </div>
                  </td>
                ))}
              </tr>
            </tbody>
          </table>
        </div>
      </Col>
    );
  }

  renderIsolationToExposureTable(data) {
    return (
      <Col xl="12">
        <div className="analytics-table-header">Moved from Isolation to Exposure</div>
        <div className="table-responsive">
          <table className="analytics-table">
            <thead>
              <tr className="g-border-bottom text-center header">
                <th></th>
                <th>Last 24h</th>
                <th>Last 7d</th>
                <th>Last 14d</th>
                <th>Cumulative</th>
              </tr>
            </thead>
            <tbody>
              <tr className="g-border-bottom">
                <td className="header">Moved from Isolation to Exposure Workflow</td>
                {data.map((x, i) => (
                  <td key={i}>
                    <div className="count-percent-container">
                      <span className="number">{x.isolation_to_exposure_total.value}</span>
                      <span className="percentage align-bottom">({x.isolation_to_exposure_total.percentage})</span>
                    </div>
                  </td>
                ))}
              </tr>
            </tbody>
          </table>
        </div>
      </Col>
    );
  }

  render() {
    return (
      <Card>
        <Card.Header as="h4" className="text-center">
          Monitoree Flow Over Time (All Records)
        </Card.Header>
        <Card.Body>
          <Row>{this.workflowTableData.map((data, index) => this.renderWorkflowTable(data, index))}</Row>
          <div className="text-center text-secondary info-text mb-4">
            <i className="fas fa-info-circle mr-1"></i>
            Cumulative includes all incoming and outgoing counts ever recorded for this jurisdiction
          </div>
          <Row>
            {this.renderExposureToCaseTable(this.exposureToCaseTableData)}
            {this.renderIsolationToExposureTable(this.isolationToExposureTableData)}
          </Row>
          <div className="text-center text-secondary info-text mb-1">
            <i className="fas fa-info-circle mr-1"></i>
            Cumulative includes only monitorees that were enrolled in the system after {RELEASE_DATE}
          </div>
        </Card.Body>
      </Card>
    );
  }
}

MonitoreeFlow.propTypes = {
  stats: PropTypes.object,
};

export default MonitoreeFlow;
