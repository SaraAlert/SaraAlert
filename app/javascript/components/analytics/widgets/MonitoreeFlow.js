import React from 'react';
import { PropTypes } from 'prop-types';
import _ from 'lodash';
import { Card } from 'react-bootstrap';

const WORKFLOWS = ['Exposure', 'Isolation'];

// Provide a separate array, as object-iteration order is not guaranteed in JS
const MONITOREE_FLOW_HEADERS = ['Last 24 Hours', 'Last 7 Days', 'Last 14 Days', 'Total'];

class MonitoreeFlow extends React.Component {
  constructor(props) {
    super(props);
    this.tableData = WORKFLOWS.map(workflow => {
      return MONITOREE_FLOW_HEADERS.map(time_frame => {
        let thisTimeFrameData = props.stats.monitoree_snapshots.find(
          monitoree_snapshot => monitoree_snapshot.status === workflow && monitoree_snapshot.time_frame === time_frame
        );
        return {
          time_frame,
          new_enrollments: thisTimeFrameData?.new_enrollments || 0,
          transferred_in: thisTimeFrameData?.transferred_in || 0,
          closed: thisTimeFrameData?.closed || 0,
          transferred_out: thisTimeFrameData?.transferred_out || 0,
        };
      });
    });
  }

  render() {
    return (
      <React.Fragment>
        <Card className="card-square text-center">
          <div className="analytics-card-header font-weight-bold h5">Monitoree Flow Over Time (All Records)</div>
          <Card.Body className="mt-4">
            <table className="analytics-table">
              <thead>
                <tr>
                  <th className="py-0"></th>
                  {MONITOREE_FLOW_HEADERS.map((monitoreeFlowHeader, index) => (
                    <th key={index}>{monitoreeFlowHeader}</th>
                  ))}
                </tr>
              </thead>
              {this.tableData.map((data, index) => (
                <tbody key={index}>
                  <tr>
                    <td className="font-weight-bold text-left p-0">
                      <u>{_.upperCase(WORKFLOWS[Number(index)])} WORKFLOW</u>
                    </td>
                  </tr>
                  <tr style={{ height: '20px' }}>
                    <td className="font-weight-bold text-left analytics-mf-subheader">
                      <u>INCOMING</u>
                    </td>
                  </tr>
                  <tr className="analytics-zebra-bg">
                    <td className="text-right">NEW ENROLLMENTS</td>
                    {data.map((x, index) => (
                      <td key={index}>{x.new_enrollments}</td>
                    ))}
                  </tr>
                  <tr>
                    <td className="text-right">TRANSFERRED IN</td>
                    {data.map((x, index) => (
                      <td key={index}>{x.transferred_in}</td>
                    ))}
                  </tr>
                  <tr style={{ height: '20px' }}>
                    <td className="font-weight-bold text-left analytics-mf-subheader">
                      <u>OUTGOING</u>
                    </td>
                  </tr>
                  <tr className="analytics-zebra-bg">
                    <td className="text-right">CLOSED</td>
                    {data.map((x, index) => (
                      <td key={index}>{x.closed}</td>
                    ))}
                  </tr>
                  <tr>
                    <td className="text-right">TRANSFERRED OUT</td>
                    {data.map((x, index) => (
                      <td key={index}>{x.transferred_out}</td>
                    ))}
                  </tr>
                </tbody>
              ))}
            </table>
            <div className="text-secondary fake-demographic-text mb-1">
              <i className="fas fa-info-circle mr-3 mt-2"></i>
              Total includes all incoming and outgoing counts ever recorded for this jurisdiction
            </div>
          </Card.Body>
        </Card>
      </React.Fragment>
    );
  }
}

MonitoreeFlow.propTypes = {
  stats: PropTypes.object,
};

export default MonitoreeFlow;
