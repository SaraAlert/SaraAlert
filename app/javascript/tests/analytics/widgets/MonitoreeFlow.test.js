import React from 'react'
import { shallow } from 'enzyme';
import MonitoreeFlow from '../../../components/analytics/widgets/MonitoreeFlow.js'
import mockMonitoreeSnapshot from '../../mocks/mockSnapshot'

const mocked_stats = {
  monitoree_snapshots: mockMonitoreeSnapshot
}

const allExpectedDomStrings = ['Last 24 Hours', 'Last 14 Days', 'Total', 'INCOMING', 'NEW ENROLLMENTS',
  'TRANSFERRED IN', 'OUTGOING', 'CLOSED', 'TRANSFERRED OUT'];

const allExpectedDomValues = {
  'total_closed': 14,
  'total_new_enrollments': 425,
  'total_transferred_in':  42,
  'total_transferred_out': 57,
  'last_14_days_closed': 10,
  'last_14_days_new_enrollments': 351,
  'last_14_days_transferred_in':  36,
  'last_14_days_transferred_out': 45,
  'last_24_hours_closed': 5,
  'last_24_hours_new_enrollments': 115,
  'last_24_hours_transferred_in':  14,
  'last_24_hours_transferred_out': 25,
}

let wrapped = shallow(<MonitoreeFlow stats={mocked_stats}/>);

describe('MonitoreeFlow', () => {
  it('Properly renders', () => {
    expect(wrapped).toMatchSnapshot();
  });
});