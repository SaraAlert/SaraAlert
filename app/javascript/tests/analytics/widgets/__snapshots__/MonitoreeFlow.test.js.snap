// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitoreeFlow Properly renders 1`] = `
<Fragment>
  <Card
    body={false}
    className="card-square text-center"
  >
    <CardHeader
      as="h5"
      className="text-left"
    >
      Monitoree Flow Over Time
    </CardHeader>
    <CardBody
      className="card-body"
    >
      <Row
        className="mx-md-5 mx-lg-0"
        noGutters={false}
      >
        <ForwardRef
          borderless={true}
          striped={true}
        >
          <thead>
            <tr>
              <th
                className="py-0"
              />
              <th
                className="py-0"
              >
                 Last 24 Hours 
              </th>
              <th
                className="py-0"
              >
                 Last 14 Days 
              </th>
              <th
                className="py-0"
              >
                 Total 
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              style={
                Object {
                  "height": "0px",
                }
              }
            />
            <tr>
              <td
                className="py-1 font-weight-bold text-left"
              >
                <u>
                  INCOMING
                </u>
                 
              </td>
            </tr>
            <tr>
              <td
                className="text-right"
              >
                NEW ENROLLMENTS
              </td>
              <td>
                115
              </td>
              <td>
                351
              </td>
              <td>
                425
              </td>
            </tr>
            <tr>
              <td
                className="text-right"
              >
                TRANSFERRED IN
              </td>
              <td>
                14
              </td>
              <td>
                36
              </td>
              <td>
                42
              </td>
            </tr>
            <tr
              style={
                Object {
                  "height": "0px",
                }
              }
            />
            <tr>
              <td
                className="py-1 font-weight-bold text-left"
              >
                <u>
                  OUTGOING
                </u>
                 
              </td>
            </tr>
            <tr
              className="pt-5"
            >
              <td
                className="text-right"
              >
                CLOSED
              </td>
              <td>
                5
              </td>
              <td>
                10
              </td>
              <td>
                14
              </td>
            </tr>
            <tr>
              <td
                className="text-right"
              >
                TRANSFERRED OUT
              </td>
              <td>
                25
              </td>
              <td>
                45
              </td>
              <td>
                57
              </td>
            </tr>
          </tbody>
        </ForwardRef>
      </Row>
    </CardBody>
  </Card>
</Fragment>
`;
